File: tests/benchmarks/lodash4/ds-diff-test/branch_compare.py
import os
import sys
import json

def check_file():
  f = open("diff.result", "w")
  def out(data):
    f.write(str(data) + "\n")

  lf = open("log", "w")
  def log(data):
    lf.write(str(data) + "\n")


  def spanNotFound(alarm, span):
    log("Span " + alarm + " @ " + str(span) + " not in rev_info")

  merged_name = "merged"
  filename = merged_name + ".js"

  # Load analysis result
  jalangi_alarms = open("jalangi.result", "r").readlines()
  jalangi_alarms = filter(lambda l: "branch taken at" in l, jalangi_alarms)
 
  ds_jalangi_alarms = open("ds-jalangi.result", "r").readlines()
  ds_jalangi_alarms = filter(lambda l: "branch taken at" in l, ds_jalangi_alarms)
  
  ds_safe_alarms = open("ds-safe.result", "r").readlines()
  ds_safe_alarms = filter(lambda l: l.startswith(filename), ds_safe_alarms)
  
  safe_alarms = open("safe.result", "r").readlines()
  safe_alarms = filter(lambda l: l.startswith(filename), safe_alarms)

  # Gather span info
  info = json.load(open(merged_name + "_jalangi_.json"))

  def jalangi_filter(p):
    if not p[0].isdigit():
      return False
    if len(p[1]) < 5:
      return False
    kind = p[1][4]
    return (kind == 'C' or kind == 'C2')

  info = dict(filter(jalangi_filter, info.items()))
  rev_info = dict(map(lambda p: (tuple(p[1][:-1]),p[0]), info.items()))

  # Gather span by lines
  ranges_by_line = {}
  for span in rev_info:
    line = span[0]
    if line in ranges_by_line:
      ranges_by_line[line].append(span)
    else:
      ranges_by_line[line] = [span]

  def parse_jalangi_alarm(alarm):
    taken = alarm.split()[0] == "True"
    splits = alarm.split(":")
    l1 = int(splits[1])
    c1 = int(splits[2])
    l2 = int(splits[3])
    c2 = int(splits[4].split(")")[0])

    span = (l1, c1, l2, c2)

    if not (span in rev_info):
      spanNotFound(alarm, span)
    assert(span in rev_info)

    return (span, taken)

  def parse_safe_alarm(alarm):
    [pre, taken] = alarm.split(") ==> ")
    taken = taken.strip() == "True"
    ast_len = len(pre) - (pre.find("(") + 1)
    splits = pre.split()[0].split(":")
    l1 = int(splits[1])
    if len(splits) == 4:
      c1 = int(splits[2].split("-")[0])
      l2 = int(splits[2].split("-")[1])
      # c2 = int(splits[3])
    elif "-" in splits[2]:
      c1 = int(splits[2].split("-")[0])
      l2 = l1
      # c2 = int(splits[2].split("-")[1])
    else:
      c1 = int(splits[2])
      l2 = l1
      # c2 = c1
    c2 = c1 + ast_len

    span = False

    rngs = ranges_by_line[l1]
    min_dist = -1
    for (_, s, l, e) in rngs:
      if l != l2:
        continue
      dist = abs(s - c1) + abs(e - c2)
      if min_dist == -1 or min_dist > dist:
        min_dist = dist
        span = (l1, s, l2, e)

    if span == False:
      spanNotFound(alarm, (l1, c1, l2, c2))
      raise Error("Not found")

    return (span, taken)

  # Parse
  jalangi_alarms = sorted(list(map(parse_jalangi_alarm, jalangi_alarms)))
  ds_jalangi_alarms = sorted(list(map(parse_jalangi_alarm, ds_jalangi_alarms)))
  ds_safe_alarms = sorted(list(map(parse_safe_alarm, ds_safe_alarms)))
  ds_alarms = ds_jalangi_alarms + ds_safe_alarms
  safe_alarms = sorted(list(map(parse_safe_alarm, safe_alarms)))

  out("alarms that ds miss (There shouldn't be any):")
  for a in jalangi_alarms:
    if not a in ds_alarms:
      out(a)

  out("alarms that safe miss (There shouldn't be any):")
  for a in ds_alarms:
    if not a in safe_alarms:
      out(a)

  out("alarms that ds overapproximate jalangi:")
  for a in ds_alarms:
    if not a in jalangi_alarms:
      out(a)

  out("alarms that safe more overapproximate than ds:")
  for a in safe_alarms:
    if not a in ds_alarms:
      out(a)

  #out("alarms that differ:")
  #for a in ds_alarms:
  #  b = (a[0], not a[1])
  #  if b in ds_alarms:
  #    continue
  #  if b in safe_alarms:
  #    out(a)
  f.close()
  lf.close()

def test():
  pass

if __name__ == "__main__":
  check_file()

File: eval/branch_compare.js
const fs = require('fs');

// helpers
var content = '';
function println(str) { content += str + '\n'; }
function spanNotFound(alarm, span) {
  console.warn(`Span ${alarm} @ ${span} not in info`);
}
function read(filename) { return fs.readFileSync(filename).toString(); }
function readLines(filename) { return read(filename).split('\n'); }
function readJson(filename) { return JSON.parse(read(filename)); }
function write(filename) { fs.writeFileSync(filename, content); }
function is_numeric(str) { return /^\d+$/.test(str); }
function info_filter(info) {
  let out = {};
  for (let prop in info) {
    let value = info[prop];
    if (is_numeric(prop) && value.length === 5) {
      let kind = value[4];
      if (kind == 'C' || kind == 'C2') {
        let span = new Span(value[0], value[1], value[2], value[3]);
        out[span] = prop;
      }
    }
  }
  return out;
}
function Loc(l, c) {
  this.line = l;
  this.col = c;
}
function Span(sl, sc, el, ec) {
  this.start = new Loc(sl, sc);
  this.end = new Loc(el, ec);
};
Span.prototype.toString = function toString() {
  return `${this.start.line}:${this.start.col}:${this.end.line}:${this.end.col}`;
}
Span.from = function from(str) {
  let arr = str.split(':');
  return new Span(arr[0], arr[1], arr[2], arr[3]);
}
function get_line(alarm) {
  return Number(alarm.split(':')[0]) * 1000 + Number(alarm.split(':')[1]);
}
function alarm_sort(alarms) {
  return alarms.sort((x, y) => get_line(x) - get_line(y));
}

// for jalangi
function jalangi_filter(msg) { return msg.includes('branch taken at'); }
function parse_jalangi_alarm(alarm) {
  let taken = alarm.split(' ')[0] == 'True';
  let arr = alarm.split(':');
  let span = new Span(arr[1], arr[2], arr[3], arr[4].split(')')[0]);
  if (!(span in info)) spanNotFound(alarm, span);
  return `${span}-${taken}`;
}
function read_jalangi(filename) {
  return readLines(filename).filter(jalangi_filter).map(parse_jalangi_alarm);
}

// for safe
function safe_filter(msg) { return msg.startsWith('target.js:'); }
function parse_safe_alarm(alarm) {
  let [pre, taken] = alarm.split(') ==> ');
  taken = taken.trim() == 'True';
  let ast_len = pre.length - (pre.indexOf('(') + 1)
  let arr = pre.split(' ')[0].split(':');

  var l1 = arr[1];
  if (arr.length === 4) {
    var [c1, l2] = arr[2].split('-'), c2 = arr[3];
  } else if (arr[2].includes('-')) {
    var [c1, c2] = arr[2].split('-'), l2 = l1;
  } else {
    var c1 = arr[2], l2 = l1, c2 = c1;
  }
  c2 = Number(c1) + ast_len;

  let found = undefined;

  rngs = rngs_by_line[l1] || [];
  let min_dist = -1;
  for (let span of rngs) {
    let s = span.start.col;
    let l = span.end.line;
    let e = span.end.col;
    if (l !== l2) continue;
    let dist = Math.abs(s - c1) + Math.abs(e - c2);
    if (min_dist == -1 || min_dist > dist) {
      min_dist = dist;
      found = new Span(l1, s, l2, e);
    }
  }

  if (!found) spanNotFound(alarm, new Span(l1, c1, l2, c2));

  return `${found}-${taken}`;
}
function read_safe(filename) {
  return alarm_sort(readLines(filename)
    .filter(safe_filter)
    .map(parse_safe_alarm)
  );
}

// check results
function check() {
  let safe_result = ds_mode ? 'ds-safe.result' : 'safe.result';
  if (!read(safe_result).includes("[BugDetect] END")) return false;

  // read info
  info = info_filter(readJson('target_jalangi_.json'));

  // gather span by lines
  rngs_by_line = {};
  for (let key in info) {
    let span = Span.from(key);
    let line = span.start.line;
    if (line in rngs_by_line) rngs_by_line[line].push(span);
    else rngs_by_line[line] = [span];
  }

  // reaad alarms
  jalangi_alarms = read_jalangi('jalangi.result');
  jalangi_alarms = alarm_sort(Array.from(new Set([...jalangi_alarms])));

  safe_alarms = [];
  if (ds_mode) safe_alarms = safe_alarms.concat(read_jalangi('ds-jalangi.result'));
  safe_alarms = safe_alarms.concat(read_safe(safe_result));
  safe_alarms = alarm_sort(Array.from(new Set([...safe_alarms])));

  // branch coverage
  println(`JALANGI / SAFE = ${jalangi_alarms.length} / ${safe_alarms.length}`);

  // unsound
  unsound_alarms = jalangi_alarms.filter(x => !safe_alarms.includes(x));
  println(`safe miss (jalangi - safe): ${unsound_alarms.length}`);
  unsound_alarms.forEach(x => println(x));

  // over
  over_alarms = safe_alarms.filter(x => !jalangi_alarms.includes(x));
  println(`safe overapproximate (safe - jalangi): ${over_alarms.length}`)
  over_alarms.forEach(x => println(x));

  return true;
}

// ds mode check
var ds_mode = (
  process.argv.includes('-ds') ||
  process.argv.includes('--dynamic-shortcut')
);
if (check()) write('diff.result');
